Conway fee calculation

ch1bo opened the pull request:
Makes the internal wallet use the right computation depending on the current era. We support `Babbage` and `Conway` in that part of the application now by making things a bit more polymorphic on the `era`.

Co-authored by @noonio and @ffakenz 

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated
* [x] Documentation update not needed
* [x] Haddocks updated
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Not query protocol parameters in current era, but need to select

This forces us to pick an era and eventually requires the internal
wallet to handle protocol parameters polymorphically.

ch1bo committed:
Fix the min fee for ref scripts in our e2e conway genesis

This was introduced lately and our devnet config did not match the one
from preview (or mainnet)

noonio committed:
Add test to see if we can even start in Conway era

ch1bo committed:
Bring back original queryProtocolParameters

noonio committed:
Making a sum-type to handle Babbage/Conway eras

- Using the `upgradeTx` from the ledger

ch1bo committed:
Make coverFee_ polymorphic on era

This will not work as the plutus purpose does not unify Babbage and Conway.

We should be looking for a better, more generic way to update ex units
in a transaction.

ch1bo committed:
Map ErrScriptExecutionFailed errors

noonio committed:
Starting to case on era when querying pparams

ch1bo committed:
Switch conway/babbage in coverFee by upgrade/downgrading transaction

This is missing the execution unit update

ffakenz committed:
Make wallet spec compile

ch1bo committed:
Re-enable adjustRedeemers using polymorphic SpendingPurpose patterns

noonio committed:
Update the conway genesis parameters

Still investigating why the fees aren't right, even though it's using
the right pparams (Conway) now.

ch1bo committed:
Use calcMinFeeTx to calculate fees

This is higher level and resolves reference scripts itself to find the
ref script bytes (relevant in conway).

ch1bo committed:
Recalculate script integrity hash

ch1bo committed:
Recompute integrity hash in wallet after convertConwayTx

This is needed because data is serialized different in Conway than
in Babbage.

ch1bo committed:
Only query protocol params on each coverFee invocation

This also fails now with an era mismatch exception

ch1bo committed:
Update changelog

ch1bo commented:
We can verify this working correctly on our open head on `preview` too before merging this.

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -240,8 +260,9 @@ data ErrCoverFee
   | ErrNoFuelUTxOFound
   | ErrUnknownInput {input :: TxIn}
-  | ErrScriptExecutionFailed {scriptFailure :: (PlutusPurpose AsIx LedgerEra, TransactionScriptFailure LedgerEra)}
+  | ErrScriptExecutionFailed {redeemerPointer :: Text, scriptFailure :: Text}
```
Can we use some better type here than `Text`?

ch1bo replied:
I'd like to use a different too, but unreasonably hard IMO. It would make the whole type parametric by `era` or we could do an existential `redeemerPointer :: forall a. Show a => a` (or something in between).

Ultimately, we use this data constructor only to fill another constructor with `Text`. So we did not lose anything by converting to `Text` early (right now).

v0d1ch commented on hydra-node/test/Hydra/Chain/Direct/WalletSpec.hs:
```diff
@@ -370,8 +377,7 @@ genOutputsForInputs AlonzoTx{body} = do
   tx <- arbitrary
-  body <- (\x -> x & set feeTxBodyL (Coin 0)) <$> arbitrary
-  pure $ tx{body, wits = mempty}
+  pure $ tx & bodyTxL . feeTxBodyL .~ Coin 0
```
Should we also remove all tx witnesses? Not sure if arbitrary instance contains some.

ch1bo replied:
Yeah, not sure. We can also just leave this code as it was. In the end we were not able to use this for testing the coverFee with redeemers.

v0d1ch reviewed (approved):
Let's see if it works for real!

ch1bo committed:
Remove redundant build depend in hydra-cluster

ch1bo committed:
Drop always failing cover fee test in TxSpec

This test was seemingly always failing evaluation of the created tx and
not actually testing coverFee at all. Also it was quite annoying to
maintain over the years and we have better ways to create Hydra
transactions nowadays.

noonio committed:
Output line number when failing to parse persistence data

ch1bo commented on hydra-node/src/Hydra/Persistence.hs:
```diff
@@ -90,7 +91,9 @@ createPersistenceIncremental fp = do
+                    case Aeson.eitherDecodeStrict' o of
+                      Left e -> throwIO $ PersistenceException ("Error at line: " <> show (i + 1) <> " in file " <> fp <> " - " <> e)
+                      Right decoded -> pure decoded
+               in iforM (C8.lines bs) loadOrFail
```
@noonio Why use `iforM` instead of `forM` (that is already in our prelude)?

noonio replied:
I wanted the line number in the error message; that's the only thing that has changed here.

ch1bo replied:
So we can keep `forM` and not need the import?

noonio commented on .gitignore:
```diff
@@ -35,7 +35,7 @@ hspec-results.md
 
 # demo
-devnet
+/devnet
```
fyi, this was an important fix because this banned a "devnet" folder _anywhere_; not just at the root.
